lib/memo_wise.rb

```diff
@@ -163,25 +236,46 @@ def self.has_block_arg?(method) # rubocop:disable Naming/PredicateName
   def self.prepended(target) # rubocop:disable Metrics/PerceivedComplexity
     class << target
       # NOTE: See YARD docs for {.memo_wise} directly below this method!
-      def memo_wise(method_name) # rubocop:disable Metrics/PerceivedComplexity
+      def memo_wise(method_name_or_hash) # rubocop:disable Metrics/PerceivedComplexity
+        case method_name_or_hash
+        when Symbol
+          method_name = method_name_or_hash
+        when Hash
+          unless method_name_or_hash.keys == [:self]
+            raise ArgumentError,
+                  "`:self` is the only key allowed in memo_wise"
+          end
+
+          method_name = method_name_or_hash[:self]
+        end
+
+        # In Ruby, "class methods" are implemented as normal instance methods
+        # on the "singleton class" of a given Class object, found via
+        # {Class#singleton_class}.
+        # See: https://medium.com/@leo_hetsch/demystifying-singleton-classes-in-ruby-caf3fa4c9d91
+        klass = self
+        singleton = klass.singleton_class
+
+        if singleton.private_instance_methods.include?(method_name) ||
+           singleton.instance_methods.include?(method_name)
+          MemoWise.create_memo_wise_state!(self)
+          klass = singleton
+        elsif klass.singleton_class?
+          MemoWise.create_memo_wise_state!(
+            MemoWise.original_class_from_singleton(klass)
+          )
+        end
```

Comment:
@jemmaissroff Can we re-work and clarify the branching paths of logic so that we first take into account whether the `self:` keyword arg was used? I don't think we should be following any fallback logic anymore?

Reply:
In other words, in this code, are we still looking for a singleton method first even if `self` was not passed?